Sources/_FoundationInternals/String/BidirectionalCollection.swift

```diff
@@ -12,19 +12,16 @@
 
 extension BidirectionalCollection where Index == String.Index {
     internal func _alignIndex(roundingDown i: Index) -> Index {
-        return i < endIndex ? index(before: index(after: i)) : i
+        index(i, offsetBy: 0)
     }
 
     internal func _alignIndex(roundingUp i: Index) -> Index {
         let truncated = _alignIndex(roundingDown: i)
-        if i > truncated && i < endIndex {
-            return index(after: i)
-        } else {
-            return i
-        }
+        guard i > truncated && truncated < endIndex else { return truncated }
+        return index(after: truncated)
```

Comment:
This change ensures we never return `i` as is without giving the underlying collection a chance to properly align it (and set performance flags etc.), like in point 2 above.